### PR TITLE
implemented the la_consensusState api

### PR DESCRIPTION
**Pull request by dushyantb**

### CODE REVIEW

File: src/Lachain.Core/Consensus/ConsensusManager.cs [hunk 1]
```diff
@@ -326,5 +326,10 @@ public void Terminate()
         {
             _terminated = true;
         }
+
+        public EraBroadcaster GetEras(long era)
```
Comment (sgladkov) [inline src/Lachain.Core/Consensus/ConsensusManager.cs:330]: I think it is a misleading name. Use GetEraBroadcaster  or smth like this please

File: src/Lachain.Core/RPC/HTTP/Web3/BlockchainServiceWeb3.cs [hunk 2]
```diff
@@ -668,6 +672,32 @@ private JArray GetCurrentValidators()
             }
             return validators;
         }
+        
+        [JsonRpcMethod("la_consensusState")]
+        private JArray GetConsensusState(long era)
```
Comment (sgladkov) [inline src/Lachain.Core/RPC/HTTP/Web3/BlockchainServiceWeb3.cs:677]: This call should return all currently running protocol instances,  current implementation returns only protocols for given era and user hasn't any calls to understand what era he needs,

**Review by sgladkov — CHANGES_REQUESTED** [review]

**Comment by sgladkov** [comment]
One more issue: tests are failed to build now:
`/home/travis/build/LATOKEN/lachain/test/Lachain.CoreTest/RPC/HTTP/Web3/Web3BlockchainServiceTest.cs(104,31): error CS7036: There is no argument given that corresponds to the required formal parameter 'consensusManager' of 'BlockchainServiceWeb3.BlockchainServiceWeb3(ITransactionManager, IBlockManager, ITransactionPool, IStateManager, ISnapshotIndexRepository, INetworkManager, INodeRetrieval, ISystemContractReader, IConsensusManager)' [/home/travis/build/LATOKEN/lachain/test/Lachain.CoreTest/Lachain.CoreTest.csproj]`

**Comment by dushyantb** [comment]
> One more issue: tests are failed to build now: `/home/travis/build/LATOKEN/lachain/test/Lachain.CoreTest/RPC/HTTP/Web3/Web3BlockchainServiceTest.cs(104,31): error CS7036: There is no argument given that corresponds to the required formal parameter 'consensusManager' of 'BlockchainServiceWeb3.BlockchainServiceWeb3(ITransactionManager, IBlockManager, ITransactionPool, IStateManager, ISnapshotIndexRepository, INetworkManager, INodeRetrieval, ISystemContractReader, IConsensusManager)' [/home/travis/build/LATOKEN/lachain/test/Lachain.CoreTest/Lachain.CoreTest.csproj]`

yes... looking into that only

**Comment by dushyantb** [comment]
> Can you please rebase this to the current dev ?

it is already rebased with the current dev

**Comment by HMAshiqulIslam** [comment]
https://github.com/LATOKEN/lachain/pull/179/files
should this PR change 17 files ?

**Comment by dushyantb** [comment]
due to circular rebase, it has removed the `dev` rebase... fixed the code

File: src/Lachain.Core/Consensus/ConsensusManager.cs [hunk 3]
```diff
@@ -326,5 +326,10 @@ public void Terminate()
         {
             _terminated = true;
         }
+
+        public EraBroadcaster GetEraBroadcaster()
+        {
+            return _eras[CurrentEra];
```
Comment (HMAshiqulIslam) [inline src/Lachain.Core/Consensus/ConsensusManager.cs:332]: It's possible that currentEra is null for a long period of time during the start of the machine. You can trigger it locally like this. 
(1) Start 4 nodes. Stop node-1 immediately.
(2) Wait till other blocks generate 100 blocks. 
(3) Start node-1. Now call "la_consensusState". The rpc call will throw exceptions. 

So, it's better to check if this _eras has "CurrentEra" or not and return null if it does not exist.

File: src/Lachain.Core/RPC/HTTP/Web3/BlockchainServiceWeb3.cs [hunk 4]
```diff
@@ -668,8 +672,29 @@ private JArray GetCurrentValidators()
             }
             return validators;
         }
-
-
+        
+        [JsonRpcMethod("la_consensusState")]
+        private JArray GetConsensusState(long era)
```
Comment (HMAshiqulIslam) [inline src/Lachain.Core/RPC/HTTP/Web3/BlockchainServiceWeb3.cs:677]: Should we remove this argument ? A program does not know the current era.

File: src/Lachain.Core/Consensus/ConsensusManager.cs [hunk 5]
```diff
@@ -326,5 +326,17 @@ public void Terminate()
         {
             _terminated = true;
         }
+
+        public EraBroadcaster? GetEraBroadcaster()
+        {
```
Comment (HMAshiqulIslam) [inline src/Lachain.Core/Consensus/ConsensusManager.cs:331]: Should we lock _eras before accessing it ?

File: src/Lachain.Core/RPC/HTTP/Web3/BlockchainServiceWeb3.cs [hunk 6]
```diff
@@ -668,8 +672,29 @@ private JArray GetCurrentValidators()
...
+            
+            JArray protocols = new JArray { };
+            foreach (var id in registry.Keys)
+            {
+                if(!registry[id].Terminated)
+                {
+                    var temp = new JObject
+                    {
+                        ["protocol"] = id.ToString(),
+                        ["era"] = id.Era
```
[the remark this answers is not in the document]
Comment (dushyantb) [reply src/Lachain.Core/RPC/HTTP/Web3/BlockchainServiceWeb3.cs]: fixed

**Review by HMAshiqulIslam — APPROVED** [review]
LGTM

**Review by sgladkov — APPROVED** [review]

**Pull request merged**